### PR TITLE
Add dry types to client (and also refactored dry types on job)

### PR DESCRIPTION
**Pull request by abreckner**
Unfortunately I found that when you use String keys on a dry types hash schema, you can't specify optional parameters (it just ignores them). The DSL needs to be defined with symbols so you can put the (`?`) at the end to make it optional.

You can however use `with_key_transform(&:to_sym)` option at the end which will convert your string keys to symbols automatically.

I updated the old job schema as well to reflect what I found with the client schema.

Even thought the keys are converted to symbols, when you export them to XML they are converted back to strings in any case (`:ClientID` becomes `<ClientID></ClientID>`).

if there are missing attributes we now raise a Dry::Types::ConstraintError which will need to be handled in whomever is using this gem.

### CODE REVIEW

File: lib/xpm_ruby/client.rb [hunk 1]
```diff
@@ -49,9 +49,11 @@ def build(**args)
     end
 
     def add(access_token:, xero_tenant_id:, client:)
+      validated_client = XpmRuby::Schema::Client::Add[client]
```
Comment (Subskii) [inline lib/xpm_ruby/client.rb:52]: `Schema::Client::Add[client]`

File: spec/xpm_ruby/schema/client/add_spec.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        end
+
+        it "should raise an error on contacts" do
+          hash = { "Name" => "Joe Bloggs Consulting", "Contacts" => [{ "Contact" => { "Email" => "Joe Bloggs" } }] }
+          expect { XpmRuby::Schema::Client::Add[hash] }.to raise_error(Dry::Types::ConstraintError, /Name is missing in Hash input/)
+        end
+      end
+    end
+  end
+end
```
Comment (Subskii) [inline spec/xpm_ruby/schema/client/add_spec.rb:41]: Should be able to remove the `XpmRuby` prefix here too.

File: lib/xpm_ruby/client.rb (hunk 1 above)
Comment (abreckner) [reply lib/xpm_ruby/client.rb:52]: 👍

File: spec/xpm_ruby/schema/client/add_spec.rb (hunk 2 above)
Comment (abreckner) [reply spec/xpm_ruby/schema/client/add_spec.rb:41]: 👍

**Comment by abreckner** [comment]
@Subskii BTW it's implicit in this code, but I thought I would make it explicit that if there are missing attributes we now raise a `Dry::Types::ConstraintError` which will need to be handled in whomever is using this gem (e.g. the StaffController you made).

**Pull request merged**